rllib/utils/actor_manager.py

```diff
@@ -0,0 +1,324 @@
+from collections import defaultdict
+import logging
+from typing import Callable, List, Optional, Tuple, TypeVar
+
+import ray
+from ray.actor import ActorHandle
+from ray.exceptions import RayActorError
+from ray.util.annotations import DeveloperAPI
+
+
+logger = logging.getLogger(__name__)
+# Generic type var for foreach_* methods.
+T = TypeVar("T")
+
+
+class FaultTolerantActorManager:
```

Comment:
High level comment:

Looking at the API, I understand it mimics the one we already have for the non-flexible actor manager, but the more I'm thinking about this, the more I feel like this should replace the WorkerSet, whereas, right now, it's a parallel object to be handled by Algorithm besides the WorkerSet, even though, the WorkerSet and this manager share the same workers. This makes it very hard to understand, when I have to do something with my worker set (e.g. `add_worker`) or with my actor manager (also: e.g. `add_workers()`).

Then we could also make this manager automatically re-create the failed actors based on some flex-settings, like `min_num_actors`, and `max_num_actors` + our `ignore`/`restart` settings.

Reply:
maybe things will be clearer when I send the next PR where I update WorkerSet.

we can't make this a WorkerSet, because we have sets of actors that are not WorkerSet, namely Aggregation Workers, AlphaStar gradient workers, etc. So this needs to be a manager by itself that WorkerSet builds on top of.
We can extend this to support things like min_max_num_actors at some point, but I don't see the need right now. Things should work down to 1 actor.

Btw, I started with updating WorkerSet, exactly like your idea, and had to refactor this out at some point.

Reply:
I just read the code for workerset yesterday for the first time in details. WorkerSet is also a manager of a set of RolloutWorker with specialized functionality for RolloutWorker API. I have a feeling that the "right" approach is to create a general fault tolerant ActorManager that will replace WorkerSet?

Reply:
maybe "replace" was a heavy word. You can either have workerset use Actor manager internally or it could be a sub-class of it with special functionality in place. To emphasize the shared functionality I think sub-classing is a better design if it can be pulled of.

Reply:
Agree with @kouroshHakha , WorkerSet should probably be a sub-class of this new elastic ActorManager.

But we can do this later once this new class is more polished and tested. We should also remove the local worker from WorkerSet. It doesn't belong there.